src/rust/src/lib.rs

```diff
@@ -1,132 +1,224 @@
+    processed_columns[0..nrow].fill(1.);
+
     column_names.push("intercept".to_string());
 
     // Iterate through columns
-    for (col_name, column) in data.iter() {
-        let (processed_column, mut new_column_names) = match column.rtype() {
+    let mut current_column = 1; // we passed the intercept
+    let mut data_iter = columns;
+    loop {
+        let (col_name, column) = if let Some(next_column) = data_iter.next() {
+            next_column
+        } else {
+            break;
+        };
+
+        match column.rtype() {
             // Note that factors match this first condition
-            Rtype::Integers => process_integer_column(&column, col_name, nrow),
-            Rtype::Doubles => process_double_column(&column, col_name, nrow),
-            Rtype::Strings => process_string_column(&column, col_name, nrow),
-            Rtype::Logicals => process_logical_column(&column, col_name, nrow),
-            _ => return Err(Error::Other(format!("Unsupported column type: {:?}", column.rtype()))),
+            Rtype::Integers if column.is_factor() => {
+                let nlevels = column.levels().unwrap().len() - 1;
+                process_factor_column(
+                    &column,
+                    col_name,
+                    nrow,
+                    &mut column_names,
+                    &mut processed_columns
+                        [(current_column * nrow)..((current_column + nlevels) * nrow)],
+                )
+            }
+            Rtype::Integers => process_integer_column(
+                &column,
+                col_name,
+                &mut column_names,
+                &mut processed_columns[(current_column * nrow)..((current_column + 1) * nrow)],
+            ),
+            Rtype::Doubles => process_double_column(
+                &column,
+                col_name,
+                &mut column_names,
+                &mut processed_columns[(current_column * nrow)..((current_column + 1) * nrow)],
+            ),
+            Rtype::Strings => process_string_column(
+                &column,
+                col_name,
+                nrow,
+                &mut column_names,
+                &mut processed_columns[(current_column * nrow)..],
+            ),
+            Rtype::Logicals => process_logical_column(
+                &column,
+                col_name,
+                &mut column_names,
+                &mut processed_columns[(current_column * nrow)..((current_column + 1) * nrow)],
+            ),
+            _ => {
+                return Err(Error::Other(format!(
+                    "Unsupported column type: {:?}",
+                    column.rtype()
+                )))
+            }
         };
 
-        processed_columns.push(processed_column);
-        column_names.append(&mut new_column_names);
+        if column.is_factor() {
+            current_column += column.levels().unwrap().len() - 1;
+        } else {
+            current_column += 1;
+        }
```

Comment:
This is the non-linear jump that is mentioned before, and the reason for `loop`.